README.md
# Extension to prevent reloading JS/CSS resources (on AJAX request)

This extension for Yii 2 prevent reload resources (on AJAX request), which already exist on client.
Similar extension for [yii 1.1.x](https://github.com/IStranger/yii-resource-smart-load).
This extension has more functionality than the Yii 2 native resource filter, [see detail](#similar-extensions-native-resource-filtration-script)

## Requirements

+ PHP 5.4.0+. 
+ YiiFramework 2.0.2+

## Features

+ Prevent reload of JS files
+ Prevent reload of JS inline blocks
+ Prevent reload of CSS files
+ Prevent reload of CSS inline blocks
+ Flexible configuration of resources disabling
+ Can disable native filtration of JS/CSS files

## Limitations

+ Increases incoming traffic (from client to server), because extension uses cookie and http headers.
This is especially important for sites with a large number of included resource files.
This can be adjusted by changing the hash method (see options of extension).
In addition it should be remembered that the size of the cookie is limited (in browser).
+ Extension does not work if the user has enabled filtering of http headers (for example, on the corporate proxy),
and browser not to accept cookies. However, we assume the probability of such events is low.

## Installation

+ **Installation via composer:** add to your composer.json file ("require" section) the following line  <code>"istranger/yii2-resource-smart-load": "*"</code>
  (see <a href="https://packagist.org/packages/istranger/yii2-resource-smart-load">packagist page</a>)
+ Add in config file (by default /config/web.php):

```php
'components' => [
    // ...
    'view' => [
        'class'           => 'istranger\rSmartLoad\View',
        'smartLoadConfig' => [
            // Hashing method for resource names (if string value),
            // see possible values: http://php.net/manual/en/function.hash.php#104987
            // Can be assigned "callable", for example: function ($str) { return hash('sha256', $str); }
            // 'hashMethod'               => 'md5', // default = 'crc32b'

            // Types of resources, that will be tracked by current extension
            // If =null, include all resource types: ['jsFile', 'cssFile', 'jsInline', 'cssInline']
            // 'resourceTypes'            => ['jsFile', 'jsInline'],  // default = ['jsFile']

            // Enable log on server and client side (debug mode)
            // 'enableLog'                => true, // default = false

            // Activate "smart" disabling of resources on all pages
            // 'activateOnAllPages'       => true, // default = true

            // List of resources, that always should be loaded on client
            // (by name, hash, or full URL)
            // 'alwaysReloadableResources' => ['bootstrap.js'],  // default = []

            // Disable native script filter 
            // (only for resource types specified in 'resourceTypes')
            // 'disableNativeScriptFilter' => false, // default = true
        ]
    ],
    // ...
],
```

## Usage

### Typical use case

By default, this extension disables reloading of js files only, css files and js/css inline blocks not tracked 
(see option **resourceTypes**). But you can (carefully!) disable other resources types. 

That is, each "tracked" resource on the page will be loaded **only once**, even if it will later be removed from this page.
Therefore, all JS callbacks **on first load** should be bind to the global containers (for example, document) 
using jQuery-method **.on()**. 

At the subsequent AJAX requests already loaded CSS inline blocks (or CSS files) may be replaced by new content, 
therefore, in case you have problems with several CSS styles you configure exclusions (see below).

### Advanced use case

For the analysis of disabled/loaded scripts is convenient to use an option **enableLog**, 
that output useful debug information in browser console: 

```php
    'enableLog' => true, // default = false
```

You can more flexible manage resource loading on certain pages using methods (see [examples](#examples)): 

- ``` Yii::$app->view->getRSmartLoad()->disableLoadedResources(array $types = null); ```
    Disables loading of resources, which **already loaded on client**. Calling this method disables loading 
    "client" resources, even if they will registered after calling this method.
- ``` Yii::$app->view->getRSmartLoad()->disableResources(array $resourceList, array $types = null); ```
    Disables loading of **given** resources. Calling this method disables loading given resources, 
    even if they will registered after calling this method. Resource list can contain:
    - for JS/CSS files: full URL, basename, or hash
    - for JS/CSS inline blocks: full content of block, or hash
    - ```array('*')``` - disables all resources
 
These methods can be invoked in any actions. The argument **$types** is an array of tracked resource types, 
that should be excluded from the page. By default (=null), tracked all types.  
Array can be defined using constants:

```php
    array(
        RSmartLoad::RESOURCE_TYPE_JS_FILE,      // = 'jsFile'
        RSmartLoad::RESOURCE_TYPE_JS_INLINE,    // = 'jsInline'
        RSmartLoad::RESOURCE_TYPE_CSS_FILE,     // = 'cssFile'
        RSmartLoad::RESOURCE_TYPE_CSS_INLINE,   // = 'cssInline'
    )
```
This restriction has higher priority than $resourceList.


In addition, you can set **activateOnAllPages = false**, and extension will be disabled on all pages. 
You will need to manually configure disabling of resources on certain pages (with the help of these methods).

Alternatively, you can configure exclusion list of resources:

```php
    'alwaysReloadableResources' => ['bootstrap.js']  // default = []
```

These resources always will be loaded on client. Each resource can be presented: 

+ resource file: as **hash**, or **full URL**, or **basename**.
+ resource inline block: as **hash**, or **resource content**.

The hash of specific resource can be get through browser console in the global object of extension (enableLog == true).
For example:
 
```javascript
yii.resourceSmartLoad.resources = {
    "89cb8371": {
        "resource": "/test_yii2/basic/web/assets/a5ccee6b/jquery.js",
        "hash": "89cb8371",
        "source": "2015-02-14 12:27:08,
            GET,
            url = http://test.dev/test_yii2/basic/web/index.php?r=site%2Fabout,
            referrer = http://test.dev/test_yii2/basic/web/index.php"
    },
    "4cc53fc1": {
        "resource": "/test_yii2/basic/web/assets/94203850/css/bootstrap.css",
        "hash": "4cc53fc1",
        "source": "2015-02-14 12:27:08,
            GET,
            url = http://test.dev/test_yii2/basic/web/index.php?r=site%2Fabout,
            referrer = http://test.dev/test_yii2/basic/web/index.php"
    },
    "b66631ac": {
        "resource": "jQuery('#w1').yiiGridView({\"filterUrl\":\"/test_yii2/basic/web/index.php?r=user%2Findex\",\"filterSelect...",
        "hash": "b66631ac",
        "source": "2015-02-14 12:27:12,
            GET/AJAX,
            url = http://test.dev/test_yii2/basic/web/index.php?r=user%2Findex,
            referrer = http://test.dev/test_yii2/basic/web/index.php?r=site%2Fabout"
    },
    "125a821d": {
        "resource": "<style>body {font-size: 110%;}</style>",
        "hash": "125a821d",
        "source": "2015-02-14 12:27:11,
            GET/AJAX,
            url = http://test.dev/test_yii2/basic/web/index.php?r=site%2Fcontact,
            referrer = http://test.dev/test_yii2/basic/web/index.php?r=site%2Fabout"
    }
}
```

### Examples

Examples of usage in controller actions (it is assumed that **activateOnAllPages = false**).

#### Disable load of all CSS inline blocks:
    
```php
Yii::$app->view->getRSmartLoad()->disableResources(['*'], [RSmartLoad::RESOURCE_TYPE_CSS_INLINE]);
```
Note: specified resources will be excluded from the page for all requests (not only AJAX).

#### Disable load of certain resources files (for all AJAX requests):

```php
if(Yii::$app->request->isAjax){
    Yii::$app->view->getRSmartLoad()->disableResources(['yii.gridView.js', 'bootstrap.css']);
}
```
Note: at normal request (not AJAX) these files will be included into page.

#### Disable load of certain resources files with restriction by type:

```php
Yii::$app->view->getRSmartLoad()->disableResources(['yii.gridView.js', 'bootstrap.css'],[
    RSmartLoad::RESOURCE_TYPE_CSS_INLINE,
    RSmartLoad::RESOURCE_TYPE_JS_FILE
]);
```
Note: will be disabled only ```'yii.gridView.js'```, because restriction by type has higher priority than $resourceList.

#### Disable load of JS inline blocks and JS files, which already exist on client:

```php
$view->getRSmartLoad()->disableLoadedResources(['*'], [
    RSmartLoad::RESOURCE_TYPE_JS_INLINE, 
    RSmartLoad::RESOURCE_TYPE_JS_FILE
]);
```
Note: resources can be disabled only on AJAX request.

#### Disable load of all resources, which already exist on client:

```php
$view->getRSmartLoad()->disableLoadedResources(['*']);
```

## Tests

Tests will be later.

## Similar extensions. Native resource filtration script

To prevent reloading scripts for Yii 1.1.x you can use [nlsclientscript](https://github.com/nlac/nlsclientscript).

Similar approach used in [Yii 2 core](https://github.com/yiisoft/yii2/blob/master/framework/assets/yii.js#L50).

However, there are a few differences:

* Extension used different algorithm: at AJAX request duplicated resource files are deleted on the client **after**
receiving the content (not on the server, as in our realization). We assume that our approach is conceptually more correct.

* Our realization don't deletes (intentionally) the resources, which included directly in html code
(without registering through View). In this case, we assume that these resources are very necessary.

* Native filter don't deletes duplicate CSS files. But there is nothing wrong, because most browsers will not reload
the files.

* Native filter cannot prevent reload of JS/CSS inline blocks.

Our extension does not interfere with native filter. They can be used together or separately.
In our extension option **disableNativeScriptFilter** can partial disable native filter (if necessary).

changelog.md
# Changelog (Resource smart load extension) 

## [v1.1.1](https://github.com/IStranger/yii2-resource-smart-load/releases/tag/v1.1.1)

- refactored and optimized client side (resource-smart-load.js)
- fix bug with unsafe cookie reading

## [v1.1.0](https://github.com/IStranger/yii2-resource-smart-load/releases/tag/v1.1.0)

- added private and public path for global js object
- refactored client side

## [v1.0.0](https://github.com/IStranger/yii2-resource-smart-load/releases/tag/v1.0.0)

- Initial release

base/RSmartLoad.php
<?php

namespace istranger\rSmartLoad\base;

/**
 * Base class RSmartLoad
 *
 * @author  G.Azamat <m@fx4web.com>
 * @link    http://fx4.ru/
 * @link    https://github.com/IStranger/yii2-resource-smart-load       Yii 2.0.x ext
 * @link    https://github.com/IStranger/yii-resource-smart-load        Yii 1.1.x ext
 */
abstract class RSmartLoad extends BaseObject
{
    /**
     * JS private path in global namespace to extension client side. Only for internal use.
     * Public path {@link RSmartLoad::jsGlobalObjPublicPath}
     */
    const JS_GLOBAL_OBJ_PRIVATE_PATH = 'window.yiiResourceSmartLoadPrivateObj';

    const RESOURCE_TYPE_JS_FILE = 'jsFile';
    const RESOURCE_TYPE_JS_INLINE = 'jsInline';
    const RESOURCE_TYPE_CSS_FILE = 'cssFile';
    const RESOURCE_TYPE_CSS_INLINE = 'cssInline';

    /**
     * @var string[] Possible types of resources
     */
    static $resourceTypesAll = array(
        self::RESOURCE_TYPE_JS_FILE,
        self::RESOURCE_TYPE_JS_INLINE,
        self::RESOURCE_TYPE_CSS_FILE,
        self::RESOURCE_TYPE_CSS_INLINE,
    );

    /**
     * @return string Name of RequestReader class
     */
    public $requestReaderClassName;

    /**
     * @var string|callable Current hashing method (for resource names).<br/>
     *                      String - name of hashing method, possible values see {@link hash_algos} and
     *                          {@link http://php.net/manual/en/function.hash.php#104987} <br/>
     *                      Callback - function, that returned string hash:
     *                      <code> function ($str) { return hash('md5', $str); } </code>
     */
    public $hashMethod = 'crc32b';

    /**
     * @var string[] Types of resources, that will be tracked by current extension.
     * If =null, include all resource types {@link $resourceTypesAll}.
     */
    public $resourceTypes = array(self::RESOURCE_TYPE_JS_FILE);

    /**
     * @var string Enables log of registered/disabled resources (on server and client side)
     */
    public $enableLog = false;

    /**
     * @var string Activates "smart" disabling of resources on all pages.
     *             You can set =false, and call method {@link disableLoadedResources} in certain controllers/actions
     */
    public $activateOnAllPages = true;

    /**
     * @var string[] List of resources, that always should be loaded on client. Each resource can be presented: <br/>
     *               - resource file: as <b>hash</b>, or <b>full URL</b>, or <b>basename</b>.<br/>
     *               - resource inline block: as <b>hash</b>, or <b>resource content</b>.
     */
    public $alwaysReloadableResources = array();

    /**
     * @var string Name of "client" variable,
     *             see detail {@link RequestReader::getClientVar}
     */
    public $clientVarName = 'resourcesList';

    private $_resourceManager;
    private $_requestReader;

    abstract protected function writeLog($msg);

    /**
     * Registers client resources of this extension and corresponding scripts.
     * Should be register jQuery + publish 'resource-smart-load.js'
     */
    abstract protected function publishExtensionResources();


    /**
     * @param IResourceManager $resourceManager        Related resource manager instance
     *                                                 (ClientScript [yii 1.x] / View [yii 2.x])
     * @param array            $config                 Configuration of current object in format:
     *                                                 ['propName' => 'propValue']
     */
    public function __construct(IResourceManager $resourceManager, $config)
    {
        $this->setProperties($config);

        $this->_resourceManager = $resourceManager;
        $this->_requestReader = new $this->requestReaderClassName;

        $this->checkProperties();
    }

    /**
     * @return string Public path, that can be used for accessing to global object (extension client side).
     *                For internal needs, use private path {@link RSmartLoad::JS_GLOBAL_OBJ_PRIVATE_PATH}.
     */
    protected function jsGlobalObjPublicPath()
    {
        return 'resourceSmartLoad';
    }

    /**
     * Checks properties values
     *
     * @throws \Exception   if some property contains incorrect value.
     */
    protected function checkProperties()
    {
        if (!$this->getRequestReader() instanceof RequestReader) {
            $this->throwException('Class for RequestReader should be inherit from "%baseClass%" ' .
                '(see SmartLoad option/property "%option%")', array(
                '%baseClass%' => RequestReader::className(),
                '%option%'    => 'requestReaderClassName'
            ));
        }

        $supportedHashAlgorithms = hash_algos();
        if (is_string($this->hashMethod) && !in_array($this->hashMethod, $supportedHashAlgorithms)) {
            $this->throwException('Incorrect hashing method (see SmartLoad option/property "%option%"). ' .
                'Supported hash algorithms: %hashAlgorithms% ', array(
                '%hashAlgorithms%' => join(', ', $supportedHashAlgorithms),
                '%option%'         => 'hashMethod'
            ));
        }

        // todo check other properties
    }

    /**
     * @return IResourceManager   Related resource manager instance
     *                            (ClientScript [yii 1.x] / View [yii 2.x])
     */
    public function getResourceManager()
    {
        return $this->_resourceManager;
    }

    /**
     * @return RequestReader   Object, that used for reading properties of current request
     */
    public function getRequestReader()
    {
        return $this->_requestReader;
    }

    public function init()
    {
        $this->publishExtensionResources();

        if ($this->activateOnAllPages) {
            $this->disableLoadedResources($this->resourceTypes);
        }
    }

    /**
     * Returns list of hashes of resources, which already loaded on client.
     * This list is sent every ajax-request in "client" variable with name {@link RSmartLoad::clientVarName}
     * (see. resourceSmartLoad.getLoadedResources() in resource-smart-load.js)
     *
     * @return string[]     List of hashes (hashed full name of the resource).
     *                      If "client" variable not found, returns empty array()
     *
     * @see RequestReader::getClientVar
     * @see resource-smart-load.js
     */
    public function getLoadedResourcesHashes()
    {
        $resourcesList = $this->getRequestReader()->getClientVar($this->clientVarName);
        return $resourcesList
            ? json_decode($resourcesList)
            : array();
    }

    /**
     * Disables loading of given resources. Resource list can contain:
     * - for JS/CSS files: full URL, basename, or hash
     * - for JS/CSS inline blocks: full content of block, or hash
     * - array('*') - disables all resources
     *
     * <b>ATTENTION!</b> Calling this method disables loading <b>given</b> resources,
     * even if they will registered after calling this method.
     *
     * @param string[] $resourceList  List of resources, that should be excluded from the page.
     * @param string[] $types         Types of resources, that should be tracked. Possible values see
     *                                {@link RSmartLoad::resourceTypesAll}. By default (=null), tracked all types.
     *                                This restriction has higher priority than $resourceList.
     *
     * @see RSmartLoad::disableLoadedResources
     */
    public function disableResources(array $resourceList, array $types = null)
    {
        $self = $this;
        $this->getResourceManager()->executeRightBeforeResourceRender(function ($resourceManager) use ($self, $types, $resourceList) {
            $incResources = $this->filterResourcesByType($resourceList, $types);
            $this->_publishExtensionClientInit();
            $this->_publishRegisteredResourcesUpdater($incResources);
        });
    }

    /**
     * Disables loading of resources, which already loaded on client. <br/>
     * Used at AJAX requests. List of resource hashes obtained from "client" variable
     * {@link RequestReader::getClientVar}.
     *
     * <b>ATTENTION!</b> Calling this method disables loading <b>"client"</b> resources,
     * even if they will registered after calling this method.
     *
     * @param string[] $types  Types of resources, that should be disabled.
     *                         Possible values see {@link resourceTypesAll}. By default (=null), tracked all types.
     *
     * @see RSmartLoad::disableResources
     */
    public function disableLoadedResources(array $types = null)
    {
        $hashList = $this->getLoadedResourcesHashes();
        $this->disableResources($hashList, $types);
    }

    /**
     * Filters resources by $excludeList ({@link filterResourcesByType}) depending on given types.
     *
     * @param string[] $excludeList  List of resources, that should be excluded.
     *                               Format of array see in corresponding specific methods.
     * @param string[] $types        Types of resources, that should be filtered.
     *                               Possible values see {@link resourceTypesAll}.
     *                               Other types will be included certainly.
     * @returns string[]             List of resources, that will be included (only for tracked $types).
     */
    protected function filterResourcesByType(array $excludeList, array $types = null)
    {
        $self = $this;
        $types = is_array($types) ? $types : static::$resourceTypesAll;

        $incResourceList = array();
        $this->getResourceManager()->resourceFilterByFn(
            function ($resourceId, $type) use ($self, $excludeList, $types, &$incResourceList) {
                $include = true;
                if (in_array($type, $types)) {
                    $include = $self->shouldBeLoaded($resourceId, $excludeList);
                    if ($include) {
                        $incResourceList[] = $resourceId;
                    }
                }
                //$self->_log(array('$resourceId' => $resourceId, '$type' => $type, '$include' => $include));
                return $include;
            });
        return $incResourceList;
    }

    /**
     * Checks, should be loaded given resource
     *
     * @param string   $resource      Resource ID: Full URL, basename, or hash of resource [for files],
     *                                Content, or hash of resource [for inline blocks]
     * @param string[] $excludeList   List of resources, that should be excluded
     * @return bool                   If =TRUE, given resource will be included in page
     */
    protected function shouldBeLoaded($resource, array $excludeList)
    {
        if (in_array('*', $excludeList)) {
            return false;
        }

        $possibleEntries = array($resource, $this->hashString($resource)); // for URLs + inline blocks
        if ($baseName = basename($resource)) { // for URLs
            $possibleEntries[] = $baseName;
        }
        return
            (count(array_intersect($possibleEntries, $this->alwaysReloadableResources)) > 0) || // is "always reloadable"
            (count(array_intersect($possibleEntries, $excludeList)) === 0);          // is not contained in $excludeList
    }

    /**
     * Hash of given string with current hash method {@link hashMethod}
     *
     * @param string $str String, that will be hashed
     * @return string       Hashed string
     */
    protected function hashString($str)
    {
        if (is_callable($this->hashMethod)) {
            return call_user_func($this->hashMethod, $str);
        } else {
            return hash($this->hashMethod, $str);
        }
    }

    /**
     * Returns prepared data on registered resources, and current request.
     * If options {@link enableLog} is TRUE, method include extended comment data.
     *
     * @param array $incResources Plain array of resources, that will be included in page (after filtration)
     *
     * @return array Array in format: [ ['resource' => ..., 'hash' => ..., 'comment' =>  ], ... ]
     */
    private function _prepareDataOnRegisteredResources(array $incResources)
    {
        // data on current request
        $self = $this;
        $requestInfo = $this->getRequestReader()->getMethod() . ($this->getRequestReader()->getIsAjax() ? '/AJAX' : '');
        $comment = $this->enableLog
            ? array(
                date('Y-m-d H:i:s'),
                $requestInfo,
                'url = ' . $this->getRequestReader()->getCurrentURL(),
                'referrer = ' . ($this->getRequestReader()->getReferrer() ?: '')
            )
            : array($requestInfo);

        // data on registered resources
        return Helper::createByFn($incResources,
            function ($key, $resource) use ($self, $comment) {
                return array($key, array(
                    'resource' => $self->_limitStr($resource, 100),
                    'hash'     => $self->hashString($resource),
                    'comment'  => join(',' . "\n", $comment)
                ));
            });
    }


    /**
     * Logs given array (to system log)
     *
     * @param array  $resources
     * @param string $msg Message for log
     */
    private function _log($resources, $msg = 'Disabled following resources:')
    {
        if ($this->enableLog) {
            $this->writeLog($msg . "\n" . var_export($resources, true));
        }
    }

    /**
     * Registers client script for initialization of client side (+ extension options export to client side)
     */
    private function _publishExtensionClientInit()
    {
        $extensionOptionsJson = json_encode(array(
            'hashMethod'                => is_string($this->hashMethod) ? $this->hashMethod : 'php callback function',
            'resourceTypes'             => $this->resourceTypes,
            'enableLog'                 => $this->enableLog,
            'activateOnAllPages'        => $this->activateOnAllPages,
            'alwaysReloadableResources' => $this->alwaysReloadableResources,
            'jsGlobalObjPublicPath'     => $this->jsGlobalObjPublicPath(),
            'clientVarName'             => $this->clientVarName,
        ));
        $this->_publishExtensionJs('%extensionObject%.initExtension(%optionsJson%); ', array(
            '%extensionObject%' => self::JS_GLOBAL_OBJ_PRIVATE_PATH,
            '%optionsJson%'     => $extensionOptionsJson,
        ));
    }

    /**
     * Publish script, that extended (on client) list of loaded resources
     *
     * @param array $incResources Plain array of resources, that will be included in page (after filtration)
     */
    private function _publishRegisteredResourcesUpdater(array $incResources)
    {
        $resourcesJson = json_encode($this->_prepareDataOnRegisteredResources($incResources));

        $this->_publishExtensionJs(
            array(
                '$(function () {',
                '    var app = %extensionObject%;',
                '    var resources = %resourcesJson%;',
                '    resources.map(function (dataObj) { ',
                '        app.addResource(dataObj.hash, dataObj.resource, dataObj.comment); ',
                '    });',
                '});',
            ),
            array(
                '%extensionObject%' => self::JS_GLOBAL_OBJ_PRIVATE_PATH,
                '%resourcesJson%'   => $resourcesJson,
            )
        );
    }

    /**
     * Wraps in js-callback and registers given js code.
     * Used <b>only</b> for publication of scripts of <b>current extension</b>
     *
     * @param string|string[] $jsScriptLines Lines (or single line) of js code (will be joined EOL)
     * @param array           $replaceParams Params, that will be replaced in js code. Format: ['%from%' => 'to']
     */
    private function _publishExtensionJs($jsScriptLines, $replaceParams = array())
    {
        if (!is_array($jsScriptLines)) {
            $jsScriptLines = array($jsScriptLines);
        }
        array_unshift($jsScriptLines, '(function ($) {   // yii-resource-smart-load extension');
        array_push($jsScriptLines, '})(jQuery);');
        $scriptCode = join("\n", $jsScriptLines);
        $scriptCode = strtr($scriptCode, $replaceParams);

        $this->getResourceManager()->addPageJs($scriptCode);
    }

    private function _limitStr($str, $maxLength = 100)
    {
        return mb_substr($str, 0, $maxLength) . ((mb_strlen($str) > $maxLength) ? '...' : '');
    }
}